modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -414,6 +413,8 @@ private CompletableFuture<AsyncSqlCursor<List<Object>>> querySingle0(
 
         InternalTransaction outerTx = context.unwrap(InternalTransaction.class);
 
+        AtomicReference<Runnable> implicitTxRollbackAction = new AtomicReference<>(() -> {});
```

Comment:
Let's try to drop the AtomicReference. If impossible, then replace with tx wrapper (as it was in prev PR version)